### PR TITLE
[CodeGen] Implement abi.encodeWithSelector and abi.encodeWithSignature

### PR DESCRIPTION
**Pull request by jacky860226**

### CODE REVIEW

**Comment by jacky860226** [comment]
我發現實作上有Bug
我把byte4 encode成byte32了
我晚點重新push

File: lib/Frontend/ASTConsumers/ASTPrinter.cpp [hunk 1]
```diff
@@ -192,6 +192,8 @@ std::string ToString(soll::CastKind CK) {
     return "<None>";
   case soll::CastKind::IntegralCast:
     return "<IntegralCast>";
+  case soll::CastKind::FixedBytesCast:
+    return "<IntegralCast>";
```
Comment (hydai) [inline lib/Frontend/ASTConsumers/ASTPrinter.cpp:196]: Should this be `<FixedBytesCast>`?

File: lib/Sema/SemaResolveType.cpp [hunk 2]
```diff
@@ -483,13 +483,44 @@ void TypeResolver::visit(CallExprType &CE) {
       case Identifier::SpecialIdentifier::abi_encodePacked:
       case Identifier::SpecialIdentifier::abi_encode:
         break;
-      case Identifier::SpecialIdentifier::abi_encodeWithSelector:
+      case Identifier::SpecialIdentifier::abi_encodeWithSignature: {
+        auto &Arg0 = CE.getRawArguments().at(0);
```
Comment (hydai) [inline lib/Sema/SemaResolveType.cpp:487]: Does the `Arg0` have a specific name in the spec?

File: lib/Sema/SemaResolveType.cpp [hunk 3]
```diff
@@ -483,13 +483,44 @@ void TypeResolver::visit(CallExprType &CE) {
...
+        Keccak256Arguments.emplace_back(std::move(Arg0));
+        auto CallKeccak256 = std::make_unique<CallExpr>(
+            SourceRange(),
+            std::move(std::make_unique<Identifier>(
+                Token(), Identifier::SpecialIdentifier::keccak256, nullptr)),
+            std::move(Keccak256Arguments));
+        CallKeccak256->setType(
+            std::make_shared<FixedBytesType>(FixedBytesType::ByteKind::B32));
+        Arg0 = Actions.CreateDummy(std::move(CallKeccak256));
+        Arg0 = Actions.CreateDummy(std::move(Arg0));
```
Comment (hydai) [inline lib/Sema/SemaResolveType.cpp:502]: Why does `Arg0` need to pass into the `CreateDummy` twice?

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/Sema/SemaResolveType.cpp (hunk 2 above)
Comment (jacky860226) [reply lib/Sema/SemaResolveType.cpp]: change `Arg0` to `Signature`

File: lib/Sema/SemaResolveType.cpp (hunk 3 above)
Comment (jacky860226) [reply lib/Sema/SemaResolveType.cpp]: Comment added

**Comment by jacky860226** [comment]
abi.encodeWithSignature(signature, ...) == abi.encodeWithSelector(bytes4(keccak256(bytes(signature))), ...);
abi.encodeWithSelector(bytes4 selector, ...) == abi.encodePacked(selector, abi.encode(...));

**Review by hydai — APPROVED** [review]
LGTM

**Pull request merged**